### ISSUE TITLE
Transient state, SPC default behavior

### ISSUE THREAD

**Issue by duianto** [syl20bnr/spacemacs#6991]
What should the default behavior be if SPC is pressed when a transient state (TS) is open?

Currently there are 3 different behaviors in different TS's.

### 1st:
In most TS's, the spacemacs root popup window opens below the TS and the TS keys become inaccessible, and now some of the spacemacs root keys closes the TS.
#### for example:
1. open the font scaling TS: SPC z x
2. open the spacemacs root popup window: SPC
3. show the paging/help keys: C-h
4. now the font scaling TS closes.

### 2nd:
In the Code Folding TS (SPC z .), the SPC key is bound to nil, and when SPC is pressed the code folding TS closes.

### 3rd: 
In the Macrostep TS, open it by for example first opening .spacemacs (SPC f e d) and pressing: SPC m d m  or  , d m, now when SPC is pressed, then the spacemacs root popup window opens below the macrostep TS but after a SPC command has been executed (ex. toggle mazimize frame: SPC T M) then the macrostep TS remains active, and it's keys keep working.

### A possible 4th option
When a TS is open, and the SPC key is pressed, then the TS would close, and the spacemacs root popup window opens.

The 1st option doesn't work with the default emacs frames size (658x640px in win 10), when SPC is pressed in the layouts TS, then this message is shown in the minibuffer:
```
Error running timer `which-key--update': (error "Window #<window 16> too small for splitting 2")
```
It's probably because both the layouts TS and the spacemacs root popup window lists a lot of keys.   

#### System Info :computer:
- OS: windows-nt
- Emacs: 24.5.1
- Spacemacs: 0.105.22
- Spacemacs branch: develop (rev. 3290c8a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(helm emacs-lisp git)
```

**Comment by bmag** [comment]
**TL;DR:** 1st instance is a which-key/hydra confilct, report upstream. 3rd instance is a foreign-keys-persistent TS, expected behavior. 2nd instance is inconsistency, `<SPC>` binding should be removed from TS. 4th instance is expected error, maybe report upstream for nicer message. We might also want to remove `<escape>` bindings from 2 other TS's.

---

The 1st instance is almost certainly a conflict between [which-key] and [hydra]. I don't know if there's a way to solve it, but it should be reported upstream to one of packages. My guess is that which-key's `C-h` is interpreted by hydra as command, and since the TS is set to exit on "foreign keys" (commands not from the hydra), the TS closes. In fact, the TS will close whenever you execute a command that's not part of the TS, e.g. `SPC 1`, `C-x h` or even just `j`.

In the 3rd instance, the MacroStep TS is set to persist on foreign keys, and that's why the TS stays alive after a `SPC` command. It treats other commands, like `C-x h` and `j`, the same. In fact, if you use MacroStep in the 1st instance instead of font scaling, the TS doesn't close after `C-h`.

In the 2nd instance, the Code Folding TS (both of them, the origami one and the default one) specifically binds `<SPC>` (why not `SPC`?) to exit. As you point out, that is incosistent with the other TS. Slightly off topic, I don't use folding much, but a use-case of "fold some code, move to another part, unfold it and resume work" sounds plausible to me. I hesitantly propose we make the Code Folding TS stay alive after running foreign keys. Feedback from Code Folding users is welcome.

In the 4th instance, the frame is just to small for both the TS and which-key windows, so which-key can't open its window. It's a real error, that could be displayed in a more user-friendly message. However you can still complete the command you started typing. For example, type `SPC`, wait for error to show, type `w /`, the command `SPC w /` is executed. You can report it upstream to which-key.

Other TS's with uncommon exit keys:
- web-mode TS: `<escape>`
- neotree TS: `?`. That's because `?` is used as a help toggle, and `q` is used
  to close the neotree window.
- ivy TS: `<escape>`. Not sure about the details, but it looks like another special case.
- ido TS: `<escape>`.
- twittering-mode TS: `?`. Another `?` help toggle.

My opinion about `SPC` default behavior in TS is that it should not be bound, thus allowing the user to enter `SPC` commands in the TS's that persist after foreign keys. For consistency's sake, I think `SPC` shouldn't be bound to exit also on non-foreign-keys-persistent TS's. I also think `<escape>` doesn't need to be bound to exit on web-mode and ido TS's, but I'm less sure about `<escape>` than about `SPC`.

**Comment by nixmaniack** [comment]
>  I hesitantly propose we make the Code Folding TS stay alive after running foreign keys. Feedback from Code Folding users is welcome.

It currently already has `:foreign-keys run` so what you propose is already the case for exactly same use case of "fold some code, move to another part, unfold it and resume work" plus what I mentioned in the other [comment](https://github.com/syl20bnr/spacemacs/issues/6992#issuecomment-244325455).

**Comment by bmag** [comment]
@nixmaniack thanks, I didn't notice that the TS already has `:foreign-keys run`. Answered on the other thread.

**Comment by TheBB** [comment]
Speaking of threads, aren't these issues similar enough to be discussed in one place?

**Comment by TheBB** [comment]
I'll take those :+1: as yes. :-)

Please continue discussion in #6992.
